{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e698d289-f120-4eac-a473-addaa97eb187",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n"
     ]
    }
   ],
   "source": [
    "#Notebook_2 Excerses\n",
    "#Calculate the % GC and % AT content in the trna sequence\n",
    "dna = 'AAGGGCTTAGCTTAATTAAAGTGGCTGATTTGCGTTCAGTTGATGCAGAGTGGGGTTTTGCAGTCCTTA'\n",
    "def GCPercentage(dna):\n",
    "    \"\"\"\n",
    "    Takes DNA sequence and return the percentage of \n",
    "    GC content\n",
    "    \"\"\"\n",
    "    C_count=dna.count('C')\n",
    "    G_count=dna.count('G')\n",
    "\n",
    "    length_dna = len(dna)\n",
    "\n",
    "    # GC Content %\n",
    "    total_gc_content = G_count + C_count\n",
    "    gcpercent = (total_gc_content / length_dna) * 100\n",
    "    return('{:.2f}%'.format(gcpercent))\n",
    "    \n",
    "def ATPercentage(dna):\n",
    "    \"\"\"\n",
    "    Takes DNA sequence and return the percentage of \n",
    "    AT content\n",
    "    \"\"\"\n",
    "    A_count=dna.count('A')\n",
    "    T_count=dna.count('T')\n",
    "\n",
    "    length_dna = len(dna)\n",
    "   \n",
    "    # AT Content %\n",
    "    total_AT_content = T_count + A_count \n",
    "    percentage_AT = (total_AT_content / length_dna) * 100\n",
    "    return('{:.2f}%'.format(percentage_AT))\n",
    "\n",
    "\n",
    "#Given the following amino acid sequence (MNKMDLVADVAEKTDLSKAKATEVIDAVFA), \n",
    "#find the first, last and the 5th amino acids in the sequence.\n",
    "amino = 'MNKMDLVADVAEKTDLSKAKATEVIDAVFA'\n",
    "\n",
    "def findAmino(amino):\n",
    "    for i,a in enumerate(amino):\n",
    "        if i == 0: # The first aa\n",
    "            print(a)\n",
    "        elif i == 4: # The Fifth aa\n",
    "            print(a)\n",
    "        elif i == (len(amino) - 1): # The last aa\n",
    "            print(a)\n",
    "            \n",
    "#The above amino acid is a bacterial restriction enzyme that\n",
    "#recognizes \"TCCGGA\". Find the first restriction site in \n",
    "#the following sequence: AAAAATCCCGAGGCGGCTATATAGGGCTCCGGAGGCGTAATATAAAA\n",
    "dna_seq = 'AAAAATCCCGAGGCGGCTATATAGGGCTCCGGAGGCGTAATATAAAA'\n",
    "site = \"TCCGGA\"\n",
    "def restrictionSite(dna_seq, site):\n",
    "    return(dna_seq.find(site))\n",
    "\n",
    "\n",
    "# Notebook_4\n",
    "#Exercise\n",
    "#Using strings, lists, tuples and dictionaries concepts, find the \n",
    "#reverse complement of AAAAATCCCGAGGCGGCTATATAGGGCTCCGGAGGCGTAATATAAAA\n",
    "#Algorithm:\n",
    "# Question1\n",
    "dna = 'AAAAATCCCGAGGCGGCTATATAGGGCTCCGGAGGCGTAATATAAAA'\n",
    "def reverseDnaStr(dna):\n",
    "    reversed_dna = dna.replace('A','t').replace('C','g').replace('G','c').replace('G','c').replace('T','a')\n",
    "    reversed_dna = dna[::-1]\n",
    "    reversed_dna = reversed_dna.upper()\n",
    "    return(reversed_dna)\n",
    "\n",
    "# list function\n",
    "def reverseDnaLst(dna_seq):\n",
    "    dna_comp = []\n",
    "    for base in dna_seq:\n",
    "        if base == ('T'):\n",
    "            dna_comp.append('A')\n",
    "        elif base == ('A'):\n",
    "            dna_comp.append('T')\n",
    "        elif base == ('G'):\n",
    "            dna_comp.append('C')\n",
    "        elif base == ('C'):\n",
    "            dna_comp.append('G')\n",
    "        else:\n",
    "            dna_comp.append(base)\n",
    "    dna_comp.reverse()\n",
    "    d = ''.join(dna_comp)\n",
    "    return(d)\n",
    "\n",
    "\n",
    "# using dictionaries\n",
    "dna = 'AAAAATCCCGAGGCGGCTATATAGGGCTCCGGAGGCGTAATATAAAAG'\n",
    "def reverseDnaDict(seq):\n",
    "    dna_dict = {'A':'T', 'C':'G', 'G':'C', 'T':'A'}\n",
    "    rev_dna = ''\n",
    "    for seq in seq:\n",
    "        rev_dna = dna_dict[seq] + rev_dna\n",
    "    return(rev_dna)\n",
    "reverseDnaDict(dna)\n",
    "\n",
    "\n",
    "# Notebook 5\n",
    "#Create a while loop that starts with x = 0 and increments \n",
    "#x until x is equal to 5. Each iteration should print to the console.\n",
    "x=0\n",
    "def increment(x):\n",
    "    while x<=5:\n",
    "        print(x)\n",
    "        x+=1\n",
    "        \n",
    "\n",
    "#Repeat the previous problem, but the loop will skip printing \n",
    "#x = 5 to the console but will print values of x from 6 to 10.\n",
    "x=0\n",
    "def skipPrint(x):\n",
    "    while x<=10:\n",
    "        if x==5:\n",
    "            x+=1\n",
    "            continue\n",
    "        print(x)\n",
    "        x+=1\n",
    "        \n",
    "#Create a for loop that prints values from 4 to 10 to the console.\n",
    "def valuePrint(a,b):\n",
    "    for x in range(a,b):\n",
    "        print(x)\n",
    "valuePrint(4,11)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de143bb5-f0fb-4886-bf2e-9a5c61721e5d",
   "metadata": {},
   "source": [
    "### Exercise\n",
    "Let's return to our earlier exercise: calculating %GC content. In this exercise:\n",
    "- Write a function `percentageGC` that calculates the GC content of a DNA sequence\n",
    "- The function should return the %GC content\n",
    "- The Function should return a message if the provided sequence is not DNA (This should be checked by a different function, called by your function)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d971cdf6-0a82-4a7e-94ba-774cef7fc007",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write a function percentageGC that calculates the GC content of a DNA sequence\n",
    "def dnaSeqeunceValidator(dna):\n",
    "    \"\"\"A fucntion that validates a DNA sequence\n",
    "    \"\"\"\n",
    "    clean_dna_sequence = ''\n",
    "    if dna:\n",
    "        for base in dna:\n",
    "            if base.upper() not in ('A','T','C','G', 'N'):\n",
    "                print(\"Invalid character '{}' at index position '{}'\".format(base, dna.index(base)))\n",
    "                return(False, 0)\n",
    "            else:\n",
    "                clean_dna_sequence += base.upper()\n",
    "    else:\n",
    "        print('Please provide a valid sequence')\n",
    "    return(True, clean_dna_sequence)         \n",
    "    \n",
    "    \n",
    "def percentageGC(sequence):\n",
    "    \"\"\"\n",
    "    Takes a sequence of DNA and calculates the percentage of GC content\n",
    "    \"\"\"\n",
    "    # Validates the DNA sequence provided\n",
    "    flag, dna_seq = dnaSeqeunceValidator(sequence)\n",
    "    if flag and dna_seq:\n",
    "        GC_percentage = (dna_seq.upper().count('G') + dna_seq.upper().count('C'))/len(dna_seq) * 100\n",
    "        return('GC = {:.2f}%'.format(GC_percentage))\n",
    "    else:\n",
    "        print('The sequence has an Invalid character(s) which is not part of a DNA sequence')\n",
    "        \n",
    "percentageGC('ttgtft')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63260c2c-a8de-4910-bc5a-b58d5649bf7c",
   "metadata": {},
   "source": [
    "## 0.1.3. Exercise 1\n",
    "Write a function the reads the file (humchr.txt) and writes to another file (gene_names.txt) a clean list of gene names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "1fe24b27-cfbc-49ab-948c-6b90d7046b2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_names():\n",
    "    with open('../Data/humchrx.txt') as in_file:\n",
    "        with open('../Data/genenames.txt', 'w') as out_file:\n",
    "            file = in_file.readlines()\n",
    "            for x,line in enumerate(file):\n",
    "                if x > 35:\n",
    "                    split_lines = line.split()\n",
    "                    field = split_lines[0]\n",
    "                    if field.startswith('-'):\n",
    "                        break\n",
    "                    out_file.write('{}\\n'.format(field))\n",
    "gene_names()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f173ff38-3c19-4b64-8f34-435e8f87af0e",
   "metadata": {},
   "source": [
    "### Exercise 3\n",
    "\n",
    "- Using the same concept, convert your script in exercise 1 to take command line arguments (input and output files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7db038d-7233-47b8-b94b-1d12d6807ca9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function is in a file called genenames.py\n",
    "import sys\n",
    "\n",
    "input_text_file = sys.argv[1]\n",
    "output_text_file = sys.argv[2]\n",
    "\n",
    "def gene_names():\n",
    "    with open(input_text_file) as in_file:\n",
    "        with open(output_text_file, 'w') as out_file:\n",
    "            file = in_file.readlines()\n",
    "            for x,line in enumerate(file):\n",
    "                if x > 35:\n",
    "                    split_lines = line.split()\n",
    "                    field = split_lines[0]\n",
    "                    if field.startswith('-'):\n",
    "                        break\n",
    "                    out_file.write('{}\\n'.format(field))\n",
    "# Calling the function\n",
    "gene_names()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b26904a-8e1c-4403-a11a-9626db5bb2e7",
   "metadata": {},
   "source": [
    "- Using a DNA sequence read from file, answer the following questions:\n",
    "    1. Show that the DNA string contains only four letters.\n",
    "    2. In the DNA string there are regions that have a repeating letter. What is the letter and length of the longest repeating region?\n",
    "    3. How many ’ATG’s are in the DNA string?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "603351b9-9787-4971-a372-8435ab1fd591",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "input_text_file = sys.argv[1]\n",
    "substring = sys.argv[2]\n",
    "\n",
    "def gene_names(input_text_file, substring):\n",
    "    with open(input_text_file) as in_file:\n",
    "        dnastr = in_file.read()\n",
    "        dnasub = dnastr.count(substring)\n",
    "        print(dnasub)\n",
    "\n",
    "        \n",
    "        \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bbb2eb4-57d5-4fad-a82e-ac2ca36280ff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
